File: src/test/java/com/merricklabs/portlandbus/handlers/GetArrivalsIntentHandlerTest.java
package com.merricklabs.portlandbus.handlers;

import com.amazon.ask.dispatcher.request.handler.HandlerInput;
import com.amazon.ask.model.Response;
import com.amazon.ask.model.Slot;
import com.amazon.ask.model.SlotConfirmationStatus;
import com.google.common.collect.ImmutableMap;
import com.merricklabs.portlandbus.PortlandBusConfig;
import com.merricklabs.portlandbus.PortlandBusIntegrationTestBase;
import com.merricklabs.portlandbus.external.trimet.models.Arrival;
import com.merricklabs.portlandbus.mocks.MockTrimetClient;
import com.merricklabs.portlandbus.testutil.HandlerInputBuilder;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import lombok.extern.slf4j.Slf4j;
import org.testng.annotations.Test;

import static com.merricklabs.portlandbus.constants.PortlandBusIntents.GET_ARRIVALS_INTENT;
import static com.merricklabs.portlandbus.testutil.TestConstants.INTEGRATION_GROUP;
import static com.merricklabs.portlandbus.testutil.TestData.USER_ID;
import static org.testng.Assert.assertTrue;

@Slf4j
public class GetArrivalsIntentHandlerTest extends PortlandBusIntegrationTestBase {

    @Test(groups = INTEGRATION_GROUP)
    public void getNextArrivals(){
        final int stopId = 10791;

        // Initialize mock trimet client
        List<Arrival> dummyArrivals = MockTrimetClient.getDummyArrivals(stopId, 5);
        MockTrimetClient mockTrimetClient = injector.getInstance(MockTrimetClient.class);
        mockTrimetClient.setArrivals(dummyArrivals);

        PortlandBusConfig.Alexa alexaConfig = injector.getInstance(PortlandBusConfig.class).getAlexa();
        HandlerInput input = HandlerInputBuilder.builder()
                .config(alexaConfig)
                .slots(buildSlots(String.valueOf(stopId), alexaConfig))
                .intentName(GET_ARRIVALS_INTENT)
                .userId(USER_ID)
                .build()
                .getHandlerInput();
        GetArrivalsIntentHandler handler = injector.getInstance(GetArrivalsIntentHandler.class);
        Optional<Response> responseOptional = handler.handle(input);
        assertTrue(responseOptional.isPresent());
        String speechText = responseOptional.get().getOutputSpeech().toString();
        assertTrue(speechText.contains("Next arrivals at stop"));
        assertTrue(speechText.contains(String.valueOf(stopId)));
        assertTrue(speechText.contains(", and bus"));
    }

    private static Map<String, Slot> buildSlots(String stopIdString, PortlandBusConfig.Alexa alexaConfig) {
        return ImmutableMap.of(
                alexaConfig.getStopIdSlot(),
                Slot.builder()
                        .withName(alexaConfig.getStopIdSlot())
                        .withConfirmationStatus(SlotConfirmationStatus.NONE)
                        .withValue(stopIdString)
                        .build()
        );
    }
}

File: src/test/kotlin/com/merricklabs/portlandbus/handlers/GetArrivalsIntentHandlerTest.kt
package com.merricklabs.portlandbus.handlers

import com.amazon.ask.dispatcher.request.handler.HandlerInput
import com.amazon.ask.model.Slot
import com.amazon.ask.model.SlotConfirmationStatus
import com.merricklabs.portlandbus.PortlandBusConfig
import com.merricklabs.portlandbus.PortlandBusIntegrationTestBase
import com.merricklabs.portlandbus.constants.PortlandBusIntents.GET_ARRIVALS_INTENT
import com.merricklabs.portlandbus.testutil.MockTrimetUtils
import com.merricklabs.portlandbus.testutil.TestConstants.INTEGRATION_GROUP
import com.merricklabs.portlandbus.testutil.TestData.STOP_ID
import com.merricklabs.portlandbus.testutil.TestData.USER_ID
import com.merricklabs.portlandbus.testutil.TestHandlerInput
import org.koin.standalone.inject
import org.mockito.Mockito
import org.testng.Assert.assertTrue
import org.testng.annotations.Test

class GetArrivalsIntentHandlerTest : PortlandBusIntegrationTestBase() {

    private val getArrivalsIntentHandler: GetArrivalsIntentHandler by inject()

    @Test(groups = [INTEGRATION_GROUP])
    fun `get next arrivals`() {
        Mockito.`when`(trimetClient.getArrivalsForStop(STOP_ID)).thenReturn(MockTrimetUtils.getDummyArrivals(STOP_ID, 5))
        val input = getValidInput(STOP_ID)
        val responseOptional = getArrivalsIntentHandler.handle(input)
        assertTrue(responseOptional.isPresent)
        val speechText = responseOptional.get().outputSpeech.toString()
        assertTrue(speechText.contains("Next arrivals at stop"))
        assertTrue(speechText.contains(STOP_ID.toString()))
        assertTrue(speechText.contains(", and bus"))
    }

    @Test(groups = [INTEGRATION_GROUP])
    fun `test with invalid input`() {
        val input = invalidInput
        val responseOptional = getArrivalsIntentHandler.handle(input)
        assertTrue(responseOptional.isPresent)
        val speechText = responseOptional.get().outputSpeech.toString()
        assertTrue(speechText.contains("Sorry, there was a problem getting arrivals for that stop."))
    }

    private fun getValidInput(stopId: Int): HandlerInput {
        val alexaConfig = config.alexa
        return TestHandlerInput.Builder()
                .slots(buildSlots(stopId.toString(), alexaConfig))
                .userId(USER_ID)
                .intentName(GET_ARRIVALS_INTENT)
                .build()
                .handlerInput
    }

    private fun buildSlots(stopIdString: String, alexaConfig: PortlandBusConfig.Alexa): Map<String, Slot> {
        return mapOf(alexaConfig.stopIdSlot to
                Slot.builder()
                        .withName(alexaConfig.stopIdSlot)
                        .withConfirmationStatus(SlotConfirmationStatus.NONE)
                        .withValue(stopIdString)
                        .build())
    }

    private val invalidInput: HandlerInput
        get() {
            val alexaConfig = config.alexa
            return TestHandlerInput.Builder()
                    .slots(buildSlots("", alexaConfig))
                    .userId(USER_ID)
                    .intentName(GET_ARRIVALS_INTENT)
                    .build()
                    .handlerInput
        }
}